src/core/include/mp-units/bits/fixed_point.h

```diff
@@ -0,0 +1,309 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#pragma once
+
+// IWYU pragma: private, include <mp-units/framework.h>
+#include <mp-units/framework/magnitude.h>
+
+#ifndef MP_UNITS_IN_MODULE_INTERFACE
+#ifdef MP_UNITS_IMPORT_STD
+import std;
+#else
+#include <bit>
+#include <concepts>
+#include <cstdint>
+#include <cstdlib>
+#include <limits>
+#include <numbers>
+#endif
+#endif
+
+namespace mp_units {
+namespace detail {
+
+// this class synthesizes a double-width integer from two base-width integers.
+template<std::integral T>
+struct double_width_int {
+  static constexpr bool is_signed = std::is_signed_v<T>;
+  static constexpr std::size_t base_width = std::numeric_limits<std::make_unsigned_t<T>>::digits;
+  static constexpr std::size_t width = 2 * base_width;
+
+  using Th = T;
+  using Tl = std::make_unsigned_t<T>;
+
+  constexpr double_width_int() = default;
+  constexpr double_width_int(const double_width_int&) = default;
+
+  constexpr double_width_int& operator=(const double_width_int&) = default;
+
+  constexpr double_width_int(Th hi_, Tl lo_) : hi(hi_), lo(lo_) {}
+
+  explicit(true) constexpr double_width_int(long double v)
+  {
+    constexpr auto scale = int_power<long double>(2, base_width);
+    constexpr auto iscale = 1.l / scale;
+    auto scaled = v * iscale;
+    hi = static_cast<Th>(scaled);
+    auto resid = (scaled - static_cast<long double>(hi));
+    if (resid < 0) {
+      --hi;
+      resid += 1;
+    }
+    lo = static_cast<Tl>(resid * scale);
+  }
+  template<std::integral U>
+    requires(is_signed || !std::is_signed_v<U>)
+  explicit(false) constexpr double_width_int(U v)
+  {
+    if constexpr (is_signed) {
+      hi = v < 0 ? Th{-1} : Th{0};
+    } else {
+      hi = 0;
+    }
+    lo = static_cast<Tl>(v);
+  }
+
+  explicit(true) constexpr operator Th() const { return static_cast<Th>(lo); }
+
+  constexpr auto operator<=>(const double_width_int&) const = default;
+
+  // calculates the double-width product of two base-size integers; this implementation requires at least one of them to
+  // be unsigned
+  static constexpr double_width_int wide_product_of(Th lhs, Tl rhs)
+  {
+    constexpr std::size_t half_width = base_width / 2;
+    constexpr Tl msk = (Tl(1) << half_width) - 1u;
+    Th l1 = lhs >> half_width;
+    Tl l0 = static_cast<Tl>(lhs) & msk;
+    Tl r1 = rhs >> half_width;
+    Tl r0 = rhs & msk;
+    Tl t00 = l0 * r0;
+    Tl t01 = l0 * r1;
+    Th t10 = l1 * static_cast<Th>(r0);
+    Th t11 = l1 * static_cast<Th>(r1);
+    Tl m = (t01 & msk) + (static_cast<Tl>(t10) & msk) + (t00 >> half_width);
+    Th o1 = t11 + static_cast<Th>(m >> half_width) + (t10 >> half_width) + static_cast<Th>(t01 >> half_width);
+    Tl o0 = (t00 & msk) | ((m & msk) << half_width);
+    return {o1, o0};
+  }
+
+  template<std::integral Rhs>
+  friend constexpr auto operator*(const double_width_int& lhs, Rhs rhs)
```

Comment:
`[[nodiscard]]` in all the functions returning something that is not extraneous information that is typically ignored.